src/Type/Php/HrtimeFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,47 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+
+class HrtimeFunctionReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'hrtime';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$arrayType = new ConstantArrayType([new ConstantIntegerType(0), new ConstantIntegerType(1)], [new IntegerType(), new IntegerType()], 2);
+		$numberType = TypeCombinator::union(new IntegerType(), new FloatType());
+
+		if (count($functionCall->args) < 1) {
+			return $arrayType;
+		}
+
+		$argType = $scope->getType($functionCall->args[0]->value);
+		$isTrueType = (new ConstantBooleanType(true))->isSuperTypeOf($argType);
+		$isFalseType = (new ConstantBooleanType(false))->isSuperTypeOf($argType);
+		$compareTypes = $isTrueType->compareTo($isFalseType);
```

Comment:
Why the `compareTo`? Why not just

```php
if ($isTrueType->yes()) {
    return $numberType;
}

if ($isFalseType->yes()) {
    return $arrayType;
}
```

Reply:
Because of probabilities. If one of them is maybe and the other is no, maybe should win. Search for other places where `compareTo` is used :)

Reply:
Also, when both of them are yes(), your code example also wouldn't work.

Reply:
I copied that from the `microtime` extension. Which, if I recall correctly, I wrote as well. And again, if I recall correctly, this was something @ondrejmirtes suggested.

Reply:
Thanks, that makes sense, so it's basically a shorter way to write sth like

```php
if ($isTrueType->yes() || ($isTrueType->maybe() && $isFalseType->no())) {
    return $numberType;
}

if ($isFalseType->yes() || ($isFalseType->maybe() && $isTrueType->no())) {
    return $arrayType;
}
```

Reply:
Yeah, something like that, although it works in even more cases - I know that something cannot be true and false at the same time (in case of bool, both values are maybe), but in some extensions, it might happen.